mlir/lib/Transforms/Utils/DialectConversion.cpp

```diff
@@ -1478,34 +1497,12 @@ Value ConversionPatternRewriterImpl::findOrBuildReplacementValue(
   }
   Value castValue = buildUnresolvedMaterialization(
       MaterializationKind::Source, computeInsertPoint(repl), value.getLoc(),
```

Comment:
`value` would indeed work as an insertion point for the caller of this function. But `buildUnresolvedMaterialization` will also store the materialization (from `repl` -> original type, not `value` -> original type) in the `mapping`. That's problematic because other parts of the dialect conversion code base are now able to look up the materialization in the `mapping`. And for them the insertion point may not be correct. Again, kind of an edge case here...

Reply:
Both in this case, and in the `remapValues` above, we could choose to *not* store the materialization value in the mapping. That would mean that we create a new materialization each time, which would mean there's some compile time overhead. (The extra ops would disappear again when running CSE.) We could do that for N:* materializations only, which are less common. Long term, I'd like to do that as part of the One-Shot Dialect Conversion effort. I'm hesitating because:

1. MemRef -> LLVM is a 1:N replacement, so N:* materializations may be more common than I'm thinking.
2. 1:N and 1:1 would not be handled in a uniform way.
3. On the One-Shot Dialect Conversion RFC, we talked about "pooling" `unrealized_conversion_cast` ops to avoid the performance overhead to create/destroy ops. That's not part of this PR (and I'd rather put that as a separate PR).

Reply:
re: Using `value` as an insertion point. I just gave this a try, but it doesn't work. One problem is with block signature conversions. E.g.: a memref is replaced with the elements that make up a memref descriptor (during MemRef -> LLVM). In that case, `computeInsertionPoint(value)` would return the old block (that is going to be erased). The materialization must be inserted in the new block.

Reply:
(also continuing from the `remapvalues` issue above https://github.com/llvm/llvm-project/pull/116524#discussion_r1896576239)

> The conversion driver generally processes ops top-to-bottom, but that is not guaranteed if the user inserts a new op in a pattern; that op can be inserted anywhere and the driver will immediately try to legalize it. (Admittedly, this is probably a quite rare edge case...)

If the user were to insert new ops wherever they want then there doesn't really exist an insertion point that works for all use-cases either way. I think it is reasonable to expect that users may only assume dominance relations that are already present in the input IR: (i.e. that all operands dominate a user operation, and all results of an operation dominates all uses). This is commonly already the case with all kinds of patterns in MLIR and is also what would make creating target materializations for operands valid if inserted right before the operation. 

> One problem with this is that there could be multiple ConversionPattern::matchAndRewrite. The insertion point could be valid for the first matchAndRewrite (which creates the materialization), but invalid for the second matchAndRewrite (which reuses the existing materialization).

Just to ensure I understand. Assuming we have a CFG such as:
```
             ┌────────────┐             
             │            │             
             │   bb0      │             
      ┌──────┼            ┼──────┐      
      │      │ defs: v_ns │      │      
      │      │            │      │      
      │      └────────────┘      │      
      │                          │      
┌─────▼─────┐             ┌──────▼─────┐
│           │             │            │
│    bb1    │             │    bb2     │
│           │             │            │
│  op(v_ns) │             │   op(v_ns) │
└───────────┘             └────────────┘
```
and `bb1` is processed first, then it could create a target materialization for an operand using `v_n`s right? If the insertion point would then placed in `bb1` then `bb2` would nevertheless try to reuse it (since it was stored in the mapping) and create a dominance violation. 
I see how this could currently go wrong if we have a naive insertion point yeah.

> Both in this case, and in remapValues above, we could choose to not store the materialization value in the mapping. That would mean that we create a new materialization each time, which would mean there's some compile time overhead.

The way I see it there are 3 metrics we can optimize for: Simplicity, compile time and number of materializations

If a target materialization is always created then:
* Simplicity is optimized
* Number of materializations is unoptimized
* Compile time I'd expect to still be fast due to not needing to compute any kind of insertion point logic
* Additional negative: This is a regression for the 1:1 case where we have optimal number of materializations, compile time and simplicity in the current implementation.

If we always compute the optimal insertion point for a given set of materialization operands then:
* Simplicity decreses due to extra insertion point logic + reuse logic
* Number of materializations is optimal
* Compile time I'd think to be higher than the first solution due to the insertion point optimization. Maybe this can be improved by having an up-to-date dominance info, but at the cost of simplicity.
* Additional positive: Uniformity between 1:1 and 1:N cases and no regressions for 1:1

Alternatively, we could do a heuristic for when all values are within the same block, and disable the optimization (including storing in the mapping) when they aren't:
* Simplicitly decreases due to both conditional optimization logic and insertion point logic. Feels more like a heuristic.
* Number of materialization is non-optimal, but better than always creating them. Optimal for the 1:1 case!
* Compile time I'd expect to be the best here
* Additional positive: No regressions in the 1:1 case (guarantees optimal materializations)

Or lastly as you suggested, reuse for 1:1, do not for 1:N. Then:
* Simplicity is relatively high as the "always materialize" option (just a few conditionals right), but conceptually lower due to non-uniformity of 1:1 and 1:N. 
* Number of materailization is unoptimized, but optimal for the 1:1 case
* Compile I'd expect to be perfect for 1:1,  fast for 1:N
* Additional positives: No regression for 1:1

All of this to me seems like a pick your poison option. I am unconvinced that there is a large correlation between compile time and number of materializations *in the current dialect conversion*. I am also not actually sure how many times we create target materializations and how often we can reuse them. 
Unless we have more data I'd therefore opt for simpler changes to avoid premature optimizations.

I like your last suggestion a lot in that sense as it avoids regressions in the 1:N case, should have good compile time (even for 1:N *probably*, more data required) but is still rather simple. It could in the future ,once we do have data, be supplemented with heuristics getting closer to the 3rd option I mentioned.

re: The pooling on the RFC, to me it sounds rather unrelated to the optimizations here

> re: Using value as an insertion point. I just gave this a try, but it doesn't work. One problem is with block signature conversions. E.g.: a memref is replaced with the elements that make up a memref descriptor (during MemRef -> LLVM). In that case, computeInsertionPoint(value) would return the old block (that is going to be erased). The materialization must be inserted in the new block.

Is there some other way logic we can use to pass a valid insertion point to `findOrBuildReplacementValue`? The `ReplaceBlockArgRewrite` would have the old block as well sadly, right? With no real way to get a corresponding insertion point to what used to be the first operation of the block?

If there isn't really suitable context we can use for finding insertion points in the case of the block argument rewrites (and resulting source materializations), then your current implementation might actually just be the simplest solution for now I'm afraid :slightly_frowning_face:

Reply:
Dominance problems can already appear within the same basic block.

> > The conversion driver generally processes ops top-to-bottom, but that is not guaranteed if the user inserts a new op in a pattern; that op can be inserted anywhere and the driver will immediately try to legalize it. (Admittedly, this is probably a quite rare edge case...)
> 
> If the user were to insert new ops wherever they want then there doesn't really exist an insertion point that works for all use-cases either way. I think it is reasonable to expect that users may only assume dominance relations that are already present in the input IR: (i.e. that all operands dominate a user operation, and all results of an operation dominates all uses). This is commonly already the case with all kinds of patterns in MLIR and is also what would make creating target materializations for operands valid if inserted right before the operation.

Here is an example:
```mlir
%orig = "original_op"()
%0 = "producer1"()
%1 = "producer2"()

"user1"(%orig)
"user2"(%orig)
```

Let's assume that `%orig` was replaced with `[%0, %1]`. Now we convert `user1` and `user2`:

- `user2` is converted first. A materialization from `[%0, %1]` to type `X` is inserted between `user1` and `user2`.
- `user1` is converted next. The driver finds the already existing materialization to type `X` and reuses it. Now we have a dominance error.

> Compile time I'd expect to still be fast due to not needing to compute any kind of insertion point logic

There is also the overhead of creating a new `unrealized_conversion_cast` op, which is a heap allocation. I have not measured this overhead myself, but that's what sparked the discussion about "pooling". (Although it may not help here...)

The overhead could be particularly large for MemRef -> LLVM. E.g.:
```mlir
func.func @foo(%m: memref<?x?xf32>) {
  memref.store ..., %m[%i, %j]
  memref.store ..., %m[%k, %l]
  ...
```

As part of target materializations, for every user of `%m` we would build an  `llvm.undef` and 6 x `llvm.insertvalue`. (Unless we CSE the `unrealized_conversion_cast` ops before they materialized; that could be another option, but we would need a filter for CSE.)